File: LeetCode/medium/Permutations.cxx
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        dfs(nums, {}, res, unordered_map<int, bool>());
        return res;        
    }
    
    void dfs(const vector<int>& nums, vector<int> cur,
             vector<vector<int>>& res, unordered_map<int, bool> used)
    {
        if (cur.size() == nums.size())
            res.push_back(cur);
        else
        {
            for (auto e : nums)
            {
                if (used.find(e) == used.end())
                {
                    auto used_copy = used;
                    used_copy[e] = true;
                    auto cur_copy = cur;
                    cur_copy.push_back(e);
                    dfs(nums, cur_copy, res, used_copy);
                }
            }
        }
    }
};

File: LeetCode/medium/Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        def generate(l):
            if len(l) == 0:
                return []
            elif len(l) == 1:
                return [l]
            else:
                res = []
                for i, x in enumerate(l):
                    newlist = [x]
                    restlist = l[:i] + l[i+1:]
                    suffixlist = generate(restlist)
                    for s in suffixlist:
                        res.append([x] + s)
                return res
        return generate(nums)
